Implemented frequency stats for habits

alythobani opened the pull request:
- Create GraphQL query get_frequency_stats that returns performance statistics for the requested habits (default: all habits)
  - Show frequency stats on page
  - Automatically update frequency stats when new habt data is entered
  - Sort habits by completion of current fragment goal, and by days left in current fragment

Along the way a couple dev tools used:
  - Added proto-repl to project.clj as a dependency
  - Imported clojure.tools.namespace.repl/refresh for easier refreshing of namespaces while developing
  - Imported clojure.test/run-tests so you can always run tests from user namespace
  - Added documentation about refreshing within lein repl

alythobani committed:
Implemented frequency stats for habits:
  - Create GraphQL query get_frequency_stats that returns performance statistics for the requested habits (default: all habits)
  - Show frequency stats on page
  - Automatically update frequency stats when new habt data is entered
  - Sort habits by completion of current fragment goal, and by days left in current fragment

Along the way a couple dev tools used:
  - Added proto-repl to project.clj as a dependency
  - Imported clojure.tools.namespace.repl/refresh for easier refreshing of namespaces while developing
  - Imported clojure.test/run-tests so you can always run tests from user namespace
  - Added documentation about refreshing within lein repl

alythobani committed:
Slightly new design, looks way better with frequency stats now

alythobani committed:
Performance stats now display on Browse and Edit History panel

alythobani committed:
Performance stats are now also updated for the Browse and Edit History panel when new habit data is entered

alythobani committed:
Habits are now further sorted by current fragment progress

alythobani committed:
Added `habit_ids` param to `get_habits`
Refactored `get_frequency_stats` to query mongodb for all requested habits in one call

alythobani committed:
Update frequency stats on page when date changes

alythobani committed:
Page now sorts habits additionally by how much is left to do of them

alythobani committed:
Fix for automatically updating frequency stats when date changes

amilner42 commented on api/resources/schema.edn:
```diff
@@ -56,16 +56,34 @@
+                                       :amount {:type (non-null Int)}}}
+
+           :habit_frequency_stats {:description "Statistics for how the user has performed with this habit."
+                                   :fields {:habit_id {:type (non-null ID)}
```
Should any other of these fields be `non-null`. Actually yes they definitely should because on the elm side you have them be required, so might as well include that in the graphql contract

amilner42 commented on api/src/api/db.clj:
```diff
@@ -16,47 +17,136 @@
+                       $setOnInsert {:date date-time, :habit_id (ObjectId. habit_id), :_id (ObjectId.)}}
+                      {:upsert true, :return-new true}))
+
+(defn get-frequency-stats
```
A bit tricky to read code with deeply nested `let`s, can you make this use more helpers and have a less nested structure. The helpers will make it more readable

amilner42 commented on api/src/api/schema.clj:
```diff
@@ -71,33 +71,37 @@
-  [context {:keys [habit_id amount date]} value]
-  (let [date-time (date-from-y-m-d-map date)]
-    (db/set-habit-data habit_id amount date-time)))
+  [context {:keys [habit_id amount date] :as all} value]
```
NIT: Are you using `habit_id` or `amount`

amilner42 commented on api/src/api/schema.clj:
```diff
@@ -71,33 +71,37 @@
   "@refer `db/get-habit-data`."
-  [context {:keys [after_date for_habit]} value]
-  (db/get-habit-data after_date for_habit))
+  [context {:keys [after_date for_habit] :as all} value]
```
NIT: No need for destructuring if you're not using it

amilner42 commented on api/src/api/schema.clj:
```diff
@@ -71,33 +71,37 @@
   "@refer `db/delete-habit`."
-  [context {:keys [habit_id]} value]
-  (db/delete-habit habit_id))
+  [context {:keys [habit_id] :as all} value]
```
NIT: dup

amilner42 commented on api/src/api/schema.clj:
```diff
@@ -71,33 +71,37 @@
   "@refer `db/set-suspend-habit`."
-  [context {:keys [habit_id, suspended]} value]
-  (db/set-suspend-habit habit_id suspended))
+  [context {:keys [habit_id suspended] :as all} value]
```
NIT: dup

amilner42 commented on api/src/api/schema.clj:
```diff
@@ -71,33 +71,37 @@
+
+(defn resolve-query-get-frequency-stats
+  "@refer `db/get-frequency-stats`."
+  [context {:keys [habit_ids current_client_date] :as all} value]
```
NIT: Dup `habit_ids`

amilner42 commented on api/src/api/util.clj:
```diff
@@ -13,6 +13,14 @@
   [{:keys [year month day]}]
   (t/date-time year month day))
 
+(defn are-datetimes-same-date
```
NIT: probably better to just use `same-date` or `datetimes-are-same-date` so it reads nicer in an `if`

amilner42 commented on web-client/src/Api.elm:
```diff
@@ -19,95 +20,170 @@ graphQLRequest query decoder url handleError handleSuccess =
+    -> (ApiError -> b)
+    -> (HabitsAndHabitDataAndFrequencyStats -> b)
+    -> Cmd b
+queryHabitsAndHabitDataAndFrequencyStats ymd url handleError handleSuccess =
```
`url`, `handleError`, and `handleSuccess` do not needed to be listed explicitly, they're passed as the last params to the function

amilner42 commented on web-client/src/Api.elm:
```diff
@@ -273,7 +349,7 @@ mutationAddHabit createHabit =
         }"""
                 |> Util.templater templateDict
     in
-    graphQLRequest queryString <| Decode.at [ "data", "add_habit" ] Habit.decodeHabit
```
Oh god no! I **hate** when `let` indents, it makes the code nested for no reason. If you use the alpha version of elm-format then it won't indent `let`s. Also in general you don't want to be changing the indentation in the same commit as the code, cause it adds tons of unrelated diffs

amilner42 commented on web-client/src/HabitUtil.elm:
```diff
@@ -0,0 +1,188 @@
+
+
+findFrequencyStatsForHabit : Habit.Habit -> List FrequencyStats.FrequencyStats -> Result String FrequencyStats.FrequencyStats
+findFrequencyStatsForHabit habit frequencyStats =
```
This function should probably just be one line, just use forward pipes:

```elm
List.filter (\stats -> stats.habitId == habitId) frequencyStats
    |> List.head
    ...
``` 

You don't need to check here for duplicate frequency stats, just take 1, I think we're safe to not worry about sending copies

amilner42 commented on web-client/src/HabitUtil.elm:
```diff
@@ -0,0 +1,188 @@
+
+{-| Returns the habits sorted by completion, urgency, and current goal progress.
+-}
+sortHabitsByCurrentFragment : List FrequencyStats.FrequencyStats -> List Habit.Habit -> List Habit.Habit
```
Use a sort function and just pass it comparators, you can see this here:

https://github.com/amilner42/code-tidbit/blob/a8a8b3b169eb7a4c929788095ae43353d6559752/frontend/src/DefaultServices/Sort.elm#L24

Also don't put the comparators inside the function, have them be their own functions, functions with huge `let` become hard to read

amilner42 commented on web-client/src/Init.elm:
```diff
@@ -14,17 +14,27 @@ import RemoteData
-    , Api.queryHabitsAndHabitData apiBaseUrl OnGetHabitsAndHabitDataFailure OnGetHabitsAndHabitDataSuccess
-    )
+    let
+        ymd =
```
NIT: No need for a `let` here

amilner42 commented on web-client/src/Models/Habit.elm:
```diff
@@ -291,22 +291,22 @@ extractCreateHabit addHabitInputData =
-            case ( name, description, unitNameSingular, unitNamePlural, frequency ) of
-                ( Just name, Just description, Just unitNameSingular, Just unitNamePlural, Just frequency ) ->
-                    Just <| CreateGoodHabit <| CreateGoodHabitRecord name description goodHabitTime unitNameSingular unitNamePlural frequency
+        case addHabitInputData.kind of
```
These difs are the problem with changing indent on the same PR instead of in it's own PR. If there happens to be a bug too it'll be harder to find it cause of the huge dif. Formatting changes on sep commits

amilner42 commented on web-client/src/Api.elm:
```diff
@@ -19,95 +20,170 @@ graphQLRequest query decoder url handleError handleSuccess =
+    -> (ApiError -> b)
+    -> (QueriedFrequencyStats -> b)
+    -> Cmd b
+queryPastFrequencyStats ymd url handleError handleSuccess =
```
NIT: Dup (`url` `handleError` etc.)

amilner42 reviewed (commented):
I'll review the elm code again when you fix the indentation, cause right now it's a PITA to figure out the difs.

Good work dude

alythobani committed:
Elm-format@exp changes to go back to original code style

alythobani committed:
Make all `habit_frequency_stats` fields non-null

alythobani committed:
Get rid of unnecessary destructuring

alythobani committed:
Refactored complicated code for get-frequency-stats into many helper functions

alythobani committed:
Fixed `days-spanned-between-datetimes` to return the correct number
of days spanned by the datetimes together, i.e. one more than the
difference in days between the two

alythobani committed:
Make extra sure that get-consecutive-datetimes ignores the time

alythobani committed:
Add test file for api.dt-util functions

alythobani committed:
Allow the current fragment to count towards more `habit_frequency_stats`
fields if it's an already-successful good habit or an already-failed bad
habit.

alythobani committed:
Added test file for api.freq-stats-util functions

alythobani committed:
Added test for `create-habit-goal-fragment`

alythobani committed:
Small refactoring

alythobani committed:
Added test for `span-of-habit-goal-fragment`

alythobani committed:
Refactored code in api.dt-util-test into a `generate-two-random-datetimes-with-days-apart` function

alythobani committed:
Refactored api.freq-stats-util-test to use `generate-two-random-datetimes-with-days-apart`

alythobani committed:
Small refactoring

alythobani committed:
Got rid of unnecessary imports

alythobani committed:
Added test for `during-habit-goal-fragment?`

alythobani committed:
Refactored code in api.freq-stats-util-test into a `random-habit-goal-fragment-with-given-dates` function

alythobani committed:
Refactored code into a `generate-random-datetime-d-days-later` function

alythobani committed:
Refactored code into a `generate-random-habit-day-record-with-given-date` function

alythobani committed:
Added test for `get-habit-data-during-fragment`

alythobani committed:
Replaced `generate-random-datetime-d-days-later` with `generate-random-datetime-on-given-date`

alythobani committed:
Edited `random-habit-goal-fragment-with-given-dates` to randomize times

alythobani committed:
Refactored out number of test check iterations for each `prop/for-all` call

alythobani committed:
Refactored `random-habit-day-record-with-given-date` into `generate-random-habit-day-record`, for more robust use

alythobani committed:
Refactor api.freq-stats-util-test to not use `days-spanned-between-datetimes` from api.dt-util

alythobani committed:
Refactored api.freq-stats-util-test to alias api.freq-stats-util instead of referring functions

alythobani committed:
Refactored api.freq-stats-util-test to alias api.dt-util-test instead of referring functions

alythobani committed:
Tiny refactor

alythobani committed:
Refactored `random-habit-goal-fragment-with-given-dates` into `generate-random-habit-goal-fragment`, for more robust use

alythobani committed:
Small refactor

alythobani committed:
Small refactor

alythobani committed:
Added test for `evaluate-habit-goal-fragment-total-done`

alythobani committed:
Tiny refactor

alythobani committed:
Added test for `evaluate-habit-goal-fragment-successful`

alythobani committed:
Tiny refactor

alythobani committed:
Tiny comment fix

alythobani committed:
Split `get-habit-start-date-test` into two tests

alythobani committed:
Made `get-habit-start-date-other-frequencies-test` more robust

alythobani committed:
Refactoring

alythobani committed:
Tiny refactor

alythobani committed:
Added `random-datetime-on-given-date` function to api.dt-util-test

alythobani committed:
Made `get-consecutive-datetimes-test` more robust

alythobani committed:
Removed unnecessary `get-consecutive-datetimes-count-test` test

alythobani committed:
Tiny refactor

alythobani committed:
Tiny refactor

alythobani committed:
Added test for `partition-datetimes-based-on-habit-goal` with a total week frequency

alythobani committed:
Added test for `partition-datetimes-based-on-habit-goal` with an every x days frequency

alythobani committed:
Converted frequency-specific tests for `partition-datetimes-based-on-habit-goal` into one test

alythobani committed:
Remove api.core-test because it currently does nothing except fail

alythobani committed:
Tiny comment change

alythobani committed:
Refactored `get-habit-goal-amount-for-datetime-specific-day-of-week-frequency-test`
to only assert one equality at a time, so that debugging is easier if it fails

alythobani committed:
Small refactoring

alythobani committed:
Added another test case for `evaluate-habit-goal-fragment-successful`

alythobani committed:
Small variable name change

alythobani committed:
Minor refactoring

alythobani committed:
Made `create-specific-day-of-week-frequency` function

alythobani committed:
Added test for `get-habit-goal-fragments` with a bad habit and specific day of week frequency

alythobani committed:
Refactoring of variable names, for clarity

alythobani committed:
Tiny line change

alythobani committed:
Added test for `get-habit-goal-fragments` with a good habit and every x days frequency

alythobani committed:
Tiny organizational commented lines

alythobani committed:
Added tests for `update-freq-stats-with-past-fragment`

alythobani committed:
Tiny refactor

alythobani committed:
Added test for `update-freq-stats-with-current-fragment` with a good habit,
successful current fragment, and total week frequency

alythobani committed:
Added test for `update-freq-stats-with-current-fragment` with a good
habit, failed current fragment, and every x days frequency

alythobani committed:
Prevent `generate-random-habit-frequency-stats` from generating a current
streak higher than the best streak

Fix some tests related to the above

alythobani committed:
Prevent generated current fragment from exceeding the fragment length

alythobani committed:
Added test for `update-freq-stats-with-current-fragment` with a bad
habit, failed current fragment, and specific day of week frequency

alythobani committed:
Added tests for false outputs of `date-geq?` and `date-leq?`

alythobani committed:
Added a test for checking that `get-consecutive-datetimes` outputs start-of-day `DateTime`s

alythobani committed:
Added another test for `get-consecutive-datetimes`

alythobani committed:
Tiny refactor

alythobani committed:
package-lock.json file created by npm after re-installing dependencies

alythobani committed:
Split `create-habit-goal-fragment-test` into two tests

alythobani committed:
Added a couple more tests for `get-habit-data-during-fragment`

alythobani committed:
Added test for `update-freq-stats-with-current-fragment` with a bad
habit, successful current fragment, and total week frequency

alythobani committed:
Tiny doc change

alythobani committed:
Added a test for the `:total_done` field of the `habit_frequency_stats`
calculated by `compute-freq-stats-from-habit-goal-fragments`

alythobani committed:
Small refactoring

alythobani committed:
Added a test for the `:total_done` field of the `habit_frequency_stats`
calculated by `get-freq-stats-for-habit`

alythobani committed:
Refactoring of custom generator names

alythobani committed:
Small refactor: removed unnecessary explicit listing of a few parameters

alythobani committed:
Small refactor: simplified `findFrequencyStatsForHabit` function

alythobani committed:
Refactored habit sorting to be more readable (refactored out a
`compareByAll` function).

alythobani committed:
Add a `:habit_has_started` field to `habit_frequency_stats` objects

alythobani committed:
Also retrieve `habit_has_started` field as `HabitHasStarted` during FrequencyStats decoding
Sort habits by whether or not they have been started, in addition to other sorting
Treat current fragment as incomplete for any habits that have not been started

alythobani committed:
Display "Start this habit!" instead of normal frequency stats when
a habit has not been started yet

alythobani commented:
Turns out I never actually commented here but I did a bunch of editing, refactoring, fixing formatting, adding tests, etc. Feel free to review the code now
